File: dg-commons/src/main/code/SocialNetwork_Example/scala/SocialNetworkStructureBuilder.scala
package SocialNetwork_Example.scala

import java.util.concurrent.atomic.AtomicInteger

import Graph.Graph
import GraphEngine.StructureBuilder
import Helpers.FileHelper

import scala.collection.immutable

/**
 * Builds all combinations of graph structures for friendships with a maximum graph size
 */
object SocialNetworkStructureBuilder extends StructureBuilder[User, UserType.UserType, UserStub, UserTypes] {
  def nodeDataTypes = new UserTypes()

  val systemTempDir = System.getProperty("java.io.tmpdir").replaceAllLiterally("\\", "/")
  val outDir = s"$systemTempDir${if (systemTempDir.endsWith("/")) "" else "/"}SocialNetworkGraphs/"
  FileHelper.ensureEmptyDirectoryExists(outDir)

  val WRITE_STRUCTURES_IN_PARALLEL = false // Having same structure be same ID helps debugging...
  val ALSO_WRITE_AS_PNG = true
  override def generateAllNodeDataTypeGraphCombinationsOfLength(length: Int): immutable.Vector[Graph[UserStub]] = {
    val graphs = super.generateAllNodeDataTypeGraphCombinationsOfLength(length)

    if (WRITE_STRUCTURES_IN_PARALLEL) {
      val i = new AtomicInteger(0)
      graphs.par.foreach(graph => {
        graph.graphId = s"S_${i.incrementAndGet()}_${graph.allNodes.size}"
        graph.writeDotFile(s"${outDir}${graph.graphId}.gv", alsoWriteAsPng = ALSO_WRITE_AS_PNG)
      })
      println(s"Wrote ${i.get} graph files in DOT format to $outDir.")
    } else {
      var i = 0
      graphs.foreach(graph => {
        i += 1
        graph.graphId = s"S_${i}_${graph.allNodes.size}"
        graph.writeDotFile(s"${outDir}${graph.graphId}.gv", alsoWriteAsPng = ALSO_WRITE_AS_PNG)
      })
      println(s"Wrote $i graph files in DOT format to $outDir.")
    }

    graphs
  }
}

File: dg-commons/src/main/code/SocialNetwork_Example/java/SocialNetworkStructureBuilder.java
package SocialNetwork_Example.java;

import Graph.Graph;
import GraphEngine.StructureBuilder;
import Helpers.FileHelper;
import scala.collection.Iterator;
import scala.collection.immutable.Vector;
import scala.NotImplementedError;

/**
 * Builds all combinations of graph structures for friendships with a maximum graph size
 */
public class SocialNetworkStructureBuilder extends StructureBuilder<User,UserTypeVal,UserStub,UserTypes> {
    // Defined as protected in the Scala abstract class, but for some reason it requires public on the Java impl.
    public UserTypes nodeDataTypes() {
        return UserTypes.getInstance();
    }

    private final String systemTempDir = System.getProperty("java.io.tmpdir").replace('\\', '/');
    private final String outDir = systemTempDir + (systemTempDir.endsWith("/") ? "" : "/") + "SocialNetworkGraphs/";

    private static final Boolean WRITE_STRUCTURES_IN_PARALLEL = false; // Same structure = same ID helps debugging.
    private static final Boolean ALSO_WRITE_AS_PNG = true;

    private static final SocialNetworkStructureBuilder ourInstance = new SocialNetworkStructureBuilder();

    public static SocialNetworkStructureBuilder getInstance() {
        return ourInstance;
    }

    private SocialNetworkStructureBuilder() {
        FileHelper.ensureDirectoryExists(outDir);
    }

    public Vector<Graph<UserStub>> generateAllNodeDataTypeGraphCombinationsOfLength(int length) {
        Vector<Graph<UserStub>> graphs = super.generateAllNodeDataTypeGraphCombinationsOfLength(length);

        if (WRITE_STRUCTURES_IN_PARALLEL) {
            // Left as an exercise to the student.
            throw new NotImplementedError();
        } else {
            int i = 0;
            for (Iterator<Graph<UserStub>> iter = graphs.toIterator(); iter.hasNext(); ) {
                Graph<UserStub> graph = iter.next();
                graph.setGraphId("S_" + ++i + "_" + graph.allNodes().size());
                graph.writeDotFile(outDir + graph.graphId() + ".gv", false, ALSO_WRITE_AS_PNG);
            }
            System.out.println("Wrote " + i + " graph files in DOT format to " + outDir + ".");
        }

        return graphs;
    }

}
